### ISSUE TITLE
in case of  my own binding complex 

### ISSUE THREAD

**Issue by rokieplayer20** [guanjq/targetdiff#12]
I used AI ligand-protein docking tool.
So now I have a ligand sdf file that have docking coordiantes.

I want to extract pocket  but the method mentioned on github is confused in this case.

How can I extract pocket in this situation that there are protein pdb file and ligand sdf file.

Finally your advice of environment is very good. 
I share yaml content.

name: targetdiff
channels:
  - pytorch
  - nvidia
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=2_kmp_llvm
  - absl-py=1.4.0=pyhd8ed1ab_0
  - aiohttp=3.8.4=py38h01eb140_1
  - aiosignal=1.3.1=pyhd8ed1ab_0
  - async-timeout=4.0.2=pyhd8ed1ab_0
  - attrs=23.1.0=pyh71513ae_1
  - blas=1.0=mkl
  - blinker=1.6.2=pyhd8ed1ab_0
  - boost=1.74.0=py38h2b96118_5
  - boost-cpp=1.74.0=h75c5d50_8
  - brotli-python=1.0.9=py38hfa26641_9
  - bzip2=1.0.8=h7b6447c_0
  - c-ares=1.19.1=hd590300_0
  - ca-certificates=2023.5.7=hbcca054_0
  - cachetools=4.2.4=pyhd8ed1ab_0
  - cairo=1.16.0=hb05425b_5
  - certifi=2023.5.7=pyhd8ed1ab_0
  - cffi=1.15.1=py38h4a40e3a_3
  - charset-normalizer=2.0.4=pyhd3eb1b0_0
  - click=8.1.3=unix_pyhd8ed1ab_2
  - cryptography=41.0.1=py38hcdda232_0
  - cuda=11.6.1=0
  - cuda-cccl=11.6.55=hf6102b2_0
  - cuda-command-line-tools=11.6.2=0
  - cuda-compiler=11.6.2=0
  - cuda-cudart=11.6.55=he381448_0
  - cuda-cudart-dev=11.6.55=h42ad0f4_0
  - cuda-cuobjdump=11.6.124=h2eeebcb_0
  - cuda-cupti=11.6.124=h86345e5_0
  - cuda-cuxxfilt=11.6.124=hecbf4f6_0
  - cuda-driver-dev=11.6.55=0
  - cuda-gdb=12.2.53=0
  - cuda-libraries=11.6.1=0
  - cuda-libraries-dev=11.6.1=0
  - cuda-memcheck=11.8.86=0
  - cuda-nsight=12.2.53=0
  - cuda-nsight-compute=12.2.0=0
  - cuda-nvcc=11.6.124=hbba6d2d_0
  - cuda-nvdisasm=12.2.53=0
  - cuda-nvml-dev=11.6.55=haa9ef22_0
  - cuda-nvprof=12.2.60=0
  - cuda-nvprune=11.6.124=he22ec0a_0
  - cuda-nvrtc=11.6.124=h020bade_0
  - cuda-nvrtc-dev=11.6.124=h249d397_0
  - cuda-nvtx=11.6.124=h0630a44_0
  - cuda-nvvp=12.2.60=0
  - cuda-runtime=11.6.1=0
  - cuda-samples=11.6.101=h8efea70_0
  - cuda-sanitizer-api=12.2.53=0
  - cuda-toolkit=11.6.1=0
  - cuda-tools=11.6.1=0
  - cuda-visual-tools=11.6.1=0
  - cycler=0.11.0=pyhd8ed1ab_0
  - easydict=1.9=py_0
  - ffmpeg=4.3=hf484d3e_0
  - fontconfig=2.14.1=hef1e5e3_0
  - freetype=2.10.4=hca18f0e_2
  - frozenlist=1.3.3=py38h0a891b7_0
  - gds-tools=1.7.0.149=0
  - giflib=5.2.1=h5eee18b_3
  - glib=2.69.1=he621ea3_2
  - gmp=6.2.1=h295c915_3
  - gnutls=3.6.15=he1e5248_0
  - google-auth=1.35.0=pyh6c4a22f_0
  - google-auth-oauthlib=1.0.0=pyhd8ed1ab_0
  - greenlet=2.0.2=py38h17151c0_1
  - grpcio=1.56.0=py38h94a1851_2
  - icu=70.1=h27087fc_0
  - idna=3.4=pyhd8ed1ab_0
  - importlib-metadata=6.7.0=pyha770c72_0
  - intel-openmp=2023.1.0=hdb19cb5_46305
  - jpeg=9e=h5eee18b_1
  - kiwisolver=1.4.4=py38h43d8883_1
  - lame=3.100=h7b6447c_0
  - lcms2=2.12=h3be6417_0
  - ld_impl_linux-64=2.38=h1181459_1
  - lerc=4.0.0=h27087fc_0
  - libabseil=20230125.3=cxx17_h59595ed_0
  - libcublas=11.9.2.110=h5e84587_0
  - libcublas-dev=11.9.2.110=h5c901ab_0
  - libcufft=10.7.1.112=hf425ae0_0
  - libcufft-dev=10.7.1.112=ha5ce4c0_0
  - libcufile=1.7.0.149=0
  - libcufile-dev=1.7.0.149=0
  - libcurand=10.3.3.53=0
  - libcurand-dev=10.3.3.53=0
  - libcusolver=11.3.4.124=h33c3c4e_0
  - libcusparse=11.7.2.124=h7538f96_0
  - libcusparse-dev=11.7.2.124=hbbe9722_0
  - libdeflate=1.17=h5eee18b_0
  - libffi=3.4.4=h6a678d5_0
  - libgcc-ng=13.1.0=he5830b7_0
  - libgrpc=1.56.0=h3905398_2
  - libiconv=1.16=h7f8727e_2
  - libidn2=2.3.4=h5eee18b_0
  - libnpp=11.6.3.124=hd2722f0_0
  - libnpp-dev=11.6.3.124=h3c42840_0
  - libnsl=2.0.0=h7f98852_0
  - libnvjpeg=11.6.2.124=hd473ad6_0
  - libnvjpeg-dev=11.6.2.124=hb5906b9_0
  - libpng=1.6.39=h5eee18b_0
  - libprotobuf=4.23.3=hd1fb520_0
  - libsqlite=3.42.0=h2797004_0
  - libstdcxx-ng=13.1.0=hfd8a6a1_0
  - libtasn1=4.19.0=h5eee18b_0
  - libtiff=4.5.0=h6adf6a1_2
  - libunistring=0.9.10=h27cfd23_0
  - libuuid=2.38.1=h0b41bf4_0
  - libwebp=1.2.4=h11a3e52_1
  - libwebp-base=1.2.4=h5eee18b_1
  - libxcb=1.15=h0b41bf4_0
  - libxml2=2.9.14=h22db469_4
  - libzlib=1.2.13=hd590300_5
  - llvm-openmp=16.0.6=h4dfa4b3_0
  - lz4-c=1.9.4=h6a678d5_0
  - markdown=3.4.3=pyhd8ed1ab_0
  - markupsafe=2.1.3=py38h01eb140_0
  - matplotlib-base=3.4.3=py38hf4fb855_1
  - mkl=2023.1.0=h6d00ec8_46342
  - mkl-service=2.4.0=py38h5eee18b_1
  - mkl_fft=1.3.6=py38h417a72b_1
  - mkl_random=1.2.2=py38h417a72b_1
  - multidict=6.0.4=py38h1de0b5d_0
  - ncurses=6.4=h6a678d5_0
  - nettle=3.7.3=hbbd107a_1
  - nsight-compute=2023.2.0.16=0
  - oauthlib=3.2.2=pyhd8ed1ab_0
  - openbabel=3.1.1=py38h3d1cf2f_4
  - openh264=2.1.1=h4ff587b_0
  - openssl=3.1.1=hd590300_1
  - pandas=2.0.3=py38h01efb38_0
  - pcre=8.45=h9c3ff4c_0
  - pillow=9.4.0=py38h6a678d5_0
  - pip=23.1.2=pyhd8ed1ab_0
  - pixman=0.40.0=h36c2ea0_0
  - protobuf=4.23.3=py38h830738e_0
  - pthread-stubs=0.4=h36c2ea0_1001
  - pyasn1=0.4.8=py_0
  - pyasn1-modules=0.2.7=py_0
  - pycairo=1.24.0=py38h1a1917b_0
  - pycparser=2.21=pyhd3eb1b0_0
  - pyjwt=2.7.0=pyhd8ed1ab_0
  - pyopenssl=23.2.0=pyhd8ed1ab_1
  - pyparsing=3.1.0=pyhd8ed1ab_0
  - pysocks=1.7.1=pyha2e5f31_6
  - python=3.8.17=he550d4f_0_cpython
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python-lmdb=1.4.1=py38h8dc9893_0
  - python-tzdata=2023.3=pyhd8ed1ab_0
  - python_abi=3.8=3_cp38
  - pytorch=1.13.1=py3.8_cuda11.6_cudnn8.3.2_0
  - pytorch-cuda=11.6=h867d48c_1
  - pytorch-mutex=1.0=cuda
  - pytz=2023.3=pyhd8ed1ab_0
  - pyu2f=0.1.5=pyhd8ed1ab_0
  - pyyaml=6.0=py38h0a891b7_5
  - rdkit=2022.03.5=py38ha829ea6_0
  - re2=2023.03.02=h8c504da_0
  - readline=8.2=h5eee18b_0
  - reportlab=3.6.12=py38h5eee18b_0
  - requests=2.31.0=pyhd8ed1ab_0
  - requests-oauthlib=1.3.1=pyhd8ed1ab_0
  - rsa=4.9=pyhd8ed1ab_0
  - setuptools=68.0.0=pyhd8ed1ab_0
  - six=1.16.0=pyh6c4a22f_0
  - sqlalchemy=2.0.18=py38h01eb140_0
  - sqlite=3.41.2=h5eee18b_0
  - tbb=2021.8.0=hdb19cb5_0
  - tensorboard=2.13.0=pyhd8ed1ab_0
  - tensorboard-data-server=0.7.0=py38h3d167d9_0
  - tk=8.6.12=h1ccaba5_0
  - torchaudio=0.13.1=py38_cu116
  - torchvision=0.14.1=py38_cu116
  - tornado=6.3.2=py38h01eb140_0
  - typing-extensions=4.7.1=hd8ed1ab_0
  - typing_extensions=4.7.1=pyha770c72_0
  - urllib3=2.0.3=pyhd8ed1ab_1
  - werkzeug=2.3.6=pyhd8ed1ab_0
  - wheel=0.40.0=pyhd8ed1ab_0
  - xorg-libxau=1.0.11=hd590300_0
  - xorg-libxdmcp=1.1.3=h7f98852_0
  - xz=5.2.6=h166bdaf_0
  - yaml=0.2.5=h7f98852_2
  - yarl=1.9.2=py38h01eb140_0
  - zipp=3.15.0=pyhd8ed1ab_0
  - zlib=1.2.13=hd590300_5
  - zstd=1.5.2=h3eb15da_6
  - pip:
      - autodocktools-py3==1.5.7.post1+7.g725082b
      - blessed==1.20.0
      - docutils==0.17.1
      - gpustat==1.1
      - jinja2==3.1.2
      - joblib==1.3.1
      - meeko==0.1.dev3
      - mmcif-pdbx==2.0.1
      - numpy==1.20.3
      - nvidia-ml-py==12.535.77
      - pdb2pqr==3.6.1
      - propka==3.5.0
      - psutil==5.9.5
      - pyg-lib==0.2.0+pt113cu116
      - scikit-learn==1.3.0
      - scipy==1.10.1
      - threadpoolctl==3.1.0
      - torch-cluster==1.6.1+pt113cu116
      - torch-geometric==2.2.0
      - torch-scatter==2.1.1+pt113cu116
      - torch-sparse==0.6.17+pt113cu116
      - torch-spline-conv==1.2.2+pt113cu116
      - tqdm==4.65.0
      - vina==1.2.2
      - wcwidth==0.2.6

**Comment by guanjq** [comment]
Hi, for extracting the pocket, you can use the following code snippet to obtain the data:


```python
from utils.data import PDBProtein, parse_sdf_file

def pdb_to_pocket_data(pdb_path, ligand_path, radius=10):
    protein = PDBProtein(pdb_path)
    ligand = parse_sdf_file(ligand_path)
    pdb_block_pocket = protein.residues_to_pdb_block(
            protein.query_residues_ligand(ligand, radius))
    pocket_dict = PDBProtein(pdb_block_pocket).to_dict_atom()
    data = ProteinLigandData.from_protein_ligand_dicts(
        protein_dict=torchify_dict(pocket_dict),
        ligand_dict={
            'element': torch.empty([0, ], dtype=torch.long),
            'pos': torch.empty([0, 3], dtype=torch.float),
            'atom_feature': torch.empty([0, 8], dtype=torch.float),
            'bond_index': torch.empty([2, 0], dtype=torch.long),
            'bond_type': torch.empty([0, ], dtype=torch.long),
        }
    )

    return data
```

**Comment by rokieplayer20** [comment]
As you teach me, I use this function.
This is the resulft 
ProteinLigandData(
  protein_element=[348],
  protein_molecule_name='pocket',
  protein_pos=[348, 3],
  protein_is_backbone=[348],
  protein_atom_name=[348],
  protein_atom_to_aa_type=[348],
  ligand_element=[0],
  ligand_pos=[0, 3],
  ligand_atom_feature=[0, 8],
  ligand_bond_index=[2, 0],
  ligand_bond_type=[0],
  ligand_nbh_list={}
)
I don't know how to use this result.

If you are okay, you could teach me.

However,
I have apply pymol to crop the protein around the center of protein mass center point.
So, I can generate sevral molecules.

Is this method okay to be used for your programming code logic?

**Comment by guanjq** [comment]
You can replace the original function `pdb_to_pocket_data` in `scripts/sample_for_pocket.py` with this new one.

**Comment by guanjq** [comment]
and change the line 73 `data = pdb_to_pocket_data(args.pdb_path)` correspondingly

**Comment by rokieplayer20** [comment]
your advice is very good!! sucessful!
and then the prior my method pymol is valid?
